### PR TITLE
Breakpoint and instrumentation support for new C# 7 language features

### PR DESCRIPTION
**Pull request by gafter**
This is a work in progress. I will send it out for review when all of the relevant issues have been addressed. The issues that will be addressed in this PR (ideally one commit per issue) are:

- [x] #14519 Breakpoints for pattern switch and the when clause
- [x] #14439 Sequence points and breakpoints for deconstruction foreach statement
- [x] #14438 Sequence points and breakpoints for expression-bodied members
- [x] #14437 Enable setting a breakpoint in a local function
- [x] #14337 Add support for /instrument codecoverage for new language features

### CODE REVIEW

**Comment by tmat** [comment]
👍 so far. I suggest merging now and opening new PR for the rest of the changes. Although related they are of different nature.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler @tmat @JohnHamby This PR adds sequence points, IDE breakpoint support, and instrumentation for all new language constructs. Please review.

**Comment by JohnHamby** [comment]
The changes for dynamic analysis look good to me. Thanks!!!

**Comment by gafter** [comment]
@dotnet-bot test windows_release_unit64_prtest please
@dotnet-bot test windows_eta_open_prtest please

**Comment by gafter** [comment]
@dotnet-bot test windows_release_unit64_prtest please

**Comment by gafter** [comment]
@CyrusNajmabadi Can you please review this from an IDE perspective?

File: src/Features/CSharp/Portable/EditAndContinue/BreakpointSpans.cs [hunk 1]
```diff
@@ -149,19 +149,11 @@ private static int GetEndPosition(SyntaxNodeOrToken nodeOrToken)
...
-                    return (operatorDeclaration.Body != null) ? CreateSpanForBlock(operatorDeclaration.Body, position) : operatorDeclaration.ExpressionBody?.Expression.Span;
-
                 case SyntaxKind.ConversionOperatorDeclaration:
-                    var conversionDeclaration = (ConversionOperatorDeclarationSyntax)node;
-                    return (conversionDeclaration.Body != null) ? CreateSpanForBlock(conversionDeclaration.Body, position) : conversionDeclaration.ExpressionBody?.Expression.Span;
-
                 case SyntaxKind.DestructorDeclaration:
-                    return TryCreateSpanForNode(((DestructorDeclarationSyntax)node).Body, position);
+                    var methodDeclaration = (BaseMethodDeclarationSyntax)node;
+                    return (methodDeclaration.Body != null) ? CreateSpanForBlock(methodDeclaration.Body, position) : methodDeclaration.ExpressionBody?.Expression.Span;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/EditAndContinue/BreakpointSpans.cs]: can you also add support for the expression body on a constructor?

File: src/Features/CSharp/Portable/EditAndContinue/BreakpointSpans.cs [hunk 2]
```diff
@@ -284,6 +291,12 @@ private static int GetEndPosition(SyntaxNodeOrToken nodeOrToken)
                     var groupClause = (GroupClauseSyntax)node;
                     return TryCreateSpanForNode(groupClause.GroupExpression, position);
 
+                case SyntaxKind.LocalFunctionStatement:
+                    var localFunction = (LocalFunctionStatementSyntax)node;
+                    return (localFunction.Body != null) ?
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/EditAndContinue/BreakpointSpans.cs]: 1. unnecessary parens.
2. ? and : go on the next line.

**Comment by CyrusNajmabadi** [comment]
minor nits.  otherwise LGTM>

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@MattGertz @jaredpar I forgot to get approval for this RC change before integrating. This fixes sequence points, breakpoints, and instrumentation for all new language constructs.

**Comment by jaredpar** [comment]
CC @Pilchie and @ManishJayaswal who are approvals today.

**Comment by Pilchie** [comment]
Approved (after the fact).

**Pull request merged**